File: 3/3-2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <regex>

using namespace std;

void main(int argc, char* argv[])
{
    istream &input = argc > 1 ? static_cast<istream&>(ifstream{argv[1]}) : cin;
    regex mul_regex(R"(mul\((\d+),(\d+)\))");
    regex master_regex(R"((mul\(\d+,\d+\)|do\(\)|don't\(\)))");

    int total{0}, state{1};
    string s;
    while (input >> s) {
        for (auto i = sregex_iterator(s.begin(), s.end(), master_regex); i != sregex_iterator(); ++i) {
            string match = i->str();
             if (match == "do()"s)
                state = 1;
             if (match == "don't()"s)
                state = 0;
             if (state && regex_search(match, mul_regex)) {
                 smatch m = *sregex_iterator(match.begin(), match.end(), mul_regex);
                 total += stoi(m[1]) * stoi(m[2]);
             }
        }
    }

    cout << total << endl;
}

File: 3/3-2.py
import re
import sys

if len(sys.argv) < 2:
    file = sys.stdin
else:
    file = open(sys.argv[1], 'rt')

state = 1
total = 0
for line in file:
    for match in re.findall(r"(mul\(\d+,\d+\)|do\(\)|don't\(\))", line):
        if match == 'do()':
            state = 1
        if match == "don't()":
            state = 0
        if state:
            mul = re.match(r'mul\((\d+),(\d+)\)', match)
            if mul:
                total += int(mul.group(1)) * int(mul.group(2))

print(total)
